integrations/ethers-v6/src/index.ts

```diff
@@ -0,0 +1,261 @@
+	return (async (tx: TransactionRequest) => {
+		if (tx.data) {
+			const cipher = await options.fetcher.cipher({ request });
+			tx.data = hexlify(cipher.encryptCall(tx.data));
+		}
+		return send(tx);
+	}) as C;
+}
+
+export function wrapEthersSigner<P extends Signer>(
+	upstream: P,
+	options?: SapphireWrapOptions,
+): P & EIP2696_EthereumProvider {
+	if (isWrappedSigner(upstream)) {
+		return upstream;
+	}
+
+	const filled_options = fillOptions(options);
+
+	let signer: Signer;
+	if (upstream.provider) {
+		const provider = wrapEthersProvider(upstream.provider, filled_options);
+		try {
+			signer = upstream.connect(provider);
+		} catch (e: unknown) {
+			if ((e as EthersError).code !== "UNSUPPORTED_OPERATION") throw e;
+			signer = upstream;
+		}
+	} else {
+		signer = upstream;
+	}
+
+	// The signer must have been connected to a provider
+	const request = makeEthereumRequestFnFromProvider(signer);
+	if (!request) {
+		throw new SignerHasNoProviderError(
+			"Signer must be connected to a provider!",
+		);
+	}
+
+	return makeTaggedProxyObject(
+		signer,
+		SAPPHIRE_WRAPPED_ETHERS_SIGNER,
+		filled_options,
+		{
+			sendTransaction: hookEthersSend(
+				signer.sendTransaction.bind(signer),
+				filled_options,
+				request,
+			),
+			signTransaction: hookEthersSend(
+				signer.signTransaction.bind(signer),
+				filled_options,
+				request,
+			),
+			call: hookEthersCall(signer, "call", filled_options, request),
+			estimateGas: hookEthersCall(
+				signer,
+				"estimateGas",
+				filled_options,
+				request,
+			),
+			connect(provider: Provider) {
+				const wp = signer.connect(provider);
+				return wrapEthersSigner(wp, filled_options);
+			},
+			request,
+		},
+	) as P & EIP2696_EthereumProvider;
+}
+
+// ----------------------------------------------------------------------------
+
+const SAPPHIRE_WRAPPED_ETHERS_PROVIDER = "#SAPPHIRE_WRAPPED_ETHERS_PROVIDER";
+
+function isWrappedProvider<P extends Provider>(
+	p: P,
+): p is P & EIP2696_EthereumProvider {
+	return SAPPHIRE_WRAPPED_ETHERS_PROVIDER in p;
+}
+
+export class ContractRunnerHasNoProviderError extends Error {}
+
+export function wrapEthersProvider<P extends Provider>(
+	provider: P,
+	options?: SapphireWrapOptions,
+): P & EIP2696_EthereumProvider {
+	// Already wrapped, so don't wrap it again.
+	if (isWrappedProvider(provider)) {
+		return provider;
+	}
+
+	const request = makeEthereumRequestFnFromProvider(provider);
+	if (!request) {
+		throw new Error("Couldn't make request function!");
+	}
+
+	const filled_options = fillOptions(options);
+
+	return makeTaggedProxyObject(
+		provider,
+		SAPPHIRE_WRAPPED_ETHERS_PROVIDER,
+		filled_options,
+		{
+			// Calls can be unsigned, but must be enveloped.
+			call: hookEthersCall(provider, "call", filled_options, request),
+			estimateGas: hookEthersCall(
+				provider,
+				"estimateGas",
+				filled_options,
+				request,
+			),
+			request,
+			// TODO: wrap `getSigner()`
```

Comment:
Or I guess devs could do this themselves.

Reply:
So, I can't easily test this without a provider like MetaMask, as `getSigner` will try to enumerate accounts, e.g. in geth.